plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergBucketing.java

```diff
@@ -0,0 +1,219 @@
+        }
+
+        assertBucketEquals(icebergType, value);
+        assertHashEquals(icebergType, value, expectedHashValue);
+    }
+
+    private void assertBucketEquals(Type icebergType, Object value)
+    {
+        assertBucketNumberEquals(icebergType, value, Integer.MAX_VALUE);
+
+        assertBucketNumberEquals(icebergType, value, 2);
+        assertBucketNumberEquals(icebergType, value, 7);
+        assertBucketNumberEquals(icebergType, value, 31);
+        assertBucketNumberEquals(icebergType, value, 32);
+        assertBucketNumberEquals(icebergType, value, 100);
+        assertBucketNumberEquals(icebergType, value, 10000);
+        assertBucketNumberEquals(icebergType, value, 524287); // prime number
+        assertBucketNumberEquals(icebergType, value, 1 << 30);
+    }
+
+    private void assertHashEquals(Type icebergType, Object value, Integer expectedHashValue)
+    {
+        Integer icebergBucketHash = computeIcebergBucket(icebergType, value, Integer.MAX_VALUE);
+        Integer trinoBucketHash = computeTrinoBucket(icebergType, value, Integer.MAX_VALUE);
+
+        assertEquals(
+                icebergBucketHash,
+                expectedHashValue,
+                format("expected Iceberg %s(%s) hash to be %d, got %d", icebergType, value, expectedHashValue, icebergBucketHash));
+
+        assertEquals(
+                trinoBucketHash,
+                expectedHashValue,
+                format("expected Trino %s(%s) hash to be %d, got %d", icebergType, value, expectedHashValue, trinoBucketHash));
+    }
+
+    private void assertBucketNumberEquals(Type icebergType, Object value, int bucketCount)
+    {
+        Integer icebergBucket = computeIcebergBucket(icebergType, value, bucketCount);
+        Integer trinoBucket = computeTrinoBucket(icebergType, value, bucketCount);
+
+        assertEquals(
+                icebergBucket,
+                trinoBucket,
+                format("icebergType=%s, icebergBucket=%d, trinoBucket=%d", icebergType, icebergBucket, trinoBucket));
+    }
+
+    private Integer computeTrinoBucket(Type icebergType, Object value, int bucketCount)
+    {
+        io.trino.spi.type.Type trinoType = toTrinoType(icebergType, TYPE_MANAGER);
+        Function<Block, Block> bucketTransform = getBucketTransform(trinoType, bucketCount);
+
+        BlockBuilder blockBuilder = trinoType.createBlockBuilder(null, 1);
+
+        if (value != null) {
+            if (isLongDecimal(trinoType)) {
+                value = Decimals.encodeScaledValue((BigDecimal) value, ((io.trino.spi.type.DecimalType) trinoType).getScale());
+            }
+
+            if (isShortDecimal(trinoType)) {
+                value = Decimals.encodeShortScaledValue((BigDecimal) value, ((io.trino.spi.type.DecimalType) trinoType).getScale());
+            }
+
+            if (trinoType instanceof VarbinaryType) {
+                value = ((ByteBuffer) value).array();
+            }
```

Comment:
```
else {
  verify(Primitives.wrap(trinoType.getJavaType()).isinstance(value), .... );
}
```

because `writeNativeValue` doesn't do proper argument validation (https://github.com/trinodb/trino/issues/8332)